gcc/rust/checks/errors/rust-hir-pattern-analysis.h

```diff
@@ -0,0 +1,517 @@
+  virtual void visit (StructPatternFieldIdentPat &field) override;
+  virtual void visit (StructPatternFieldIdent &field) override;
+  virtual void visit (StructPattern &pattern) override;
+  virtual void visit (TupleStructItemsNoRange &tuple_items) override;
+  virtual void visit (TupleStructItemsRange &tuple_items) override;
+  virtual void visit (TupleStructPattern &pattern) override;
+  virtual void visit (TuplePatternItemsMultiple &tuple_items) override;
+  virtual void visit (TuplePatternItemsRanged &tuple_items) override;
+  virtual void visit (TuplePattern &pattern) override;
+  virtual void visit (SlicePattern &pattern) override;
+  virtual void visit (AltPattern &pattern) override;
+  virtual void visit (EmptyStmt &stmt) override;
+  virtual void visit (LetStmt &stmt) override;
+  virtual void visit (ExprStmt &stmt) override;
+  virtual void visit (TraitBound &bound) override;
+  virtual void visit (ImplTraitType &type) override;
+  virtual void visit (TraitObjectType &type) override;
+  virtual void visit (ParenthesisedType &type) override;
+  virtual void visit (ImplTraitTypeOneBound &type) override;
+  virtual void visit (TupleType &type) override;
+  virtual void visit (NeverType &type) override;
+  virtual void visit (RawPointerType &type) override;
+  virtual void visit (ReferenceType &type) override;
+  virtual void visit (ArrayType &type) override;
+  virtual void visit (SliceType &type) override;
+  virtual void visit (InferredType &type) override;
+  virtual void visit (BareFunctionType &type) override;
+};
+
+class Constructor
+{
+public:
+  enum ConstructorKind
+  {
+    // tuple or struct
+    STRUCT,
+    // enum variant
+    VARIANT,
+    // integers
+    INT_RANGE,
+    // user-provided wildcard
+    WILDCARD,
+    // references
+    REFERENCE,
+  };
+
+  static Constructor make_wildcard () { return Constructor (WILDCARD); }
+
+  static Constructor make_reference () { return Constructor (REFERENCE); }
+
+  static Constructor make_struct ()
+  {
+    Constructor c (STRUCT);
+    c.variant_idx = 0;
+    return c;
+  }
+
+  static Constructor make_variant (int variant_idx)
+  {
+    Constructor c (VARIANT);
+    c.variant_idx = variant_idx;
+    return c;
+  }
+
+  ConstructorKind get_kind () const { return kind; }
+
+  int get_variant_index () const
+  {
+    rust_assert (kind == VARIANT || kind == STRUCT);
+    return variant_idx;
+  }
+
+  bool is_covered_by (const Constructor &o) const;
+
+  bool is_wildcard () const { return kind == WILDCARD; }
+
+  // Requrired by std::set<T>
+  bool operator< (const Constructor &o) const;
+
+  std::string to_string () const;
+
+private:
+  Constructor (ConstructorKind kind) : kind (kind), variant_idx (0) {}
+  ConstructorKind kind;
+
+  union
+  {
+    // for enum variants, the variant index (always 0 for structs)
+    int variant_idx;
+
+    // for integer ranges, the range
+    struct
+    {
+      int64_t lo;
+      int64_t hi;
+    } int_range;
```

Comment:
We'll have to support more variants (floating point ranges are going away but we need to support them), we should probably use a struct with a fully qualified name in this union.